src/vector3d.cpp

```diff
@@ -0,0 +1,288 @@
+#include <iostream>
+#include <math.h>
+#include "vector3d.h"
+#include "constants.h"
+
+///////////////////////////////////////////////////////////
+// 
+// Initialize Vector3d object
+// 
+///////////////////////////////////////////////////////////
+
+// Default constructor
+Vector3d::Vector3d() {
+
+    // Initialize member variables
+
+    m_ObjectId = globalObjectId++; 
+
+    m_0 = 0.0;
+    m_1 = 0.0;
+    m_2 = 0.0;
+
+}
+
+// Initialize from values 
+
+Vector3d::Vector3d(double x, double y, double z) {
+
+    m_ObjectId = globalObjectId++; 
+
+    m_0 = x;
+    m_1 = y;
+    m_2 = z;
+}
+
+Vector3d::Vector3d(DBL_VECTOR v) {
+
+    m_ObjectId = globalObjectId++; 
+    
+    m_0 = v[0];
+    m_1 = v[1];
+    m_2 = v[2];
+}
+
+// Update after initialization
+void Vector3d::updateVector( const double x, const double y, const double z) {
+    m_0 = x;
+    m_1 = y;
+    m_2 = z;
+}
+
+///////////////////////////////
+// Overload operators
+
+// Indexing Operator
+double& Vector3d::operator[] (size_t i) {
+    switch (i) {
+        case 0: return m_0;
+        case 1: return m_1;
+        case 2: return m_2;
+        default: throw "Not an index!\n";
+    }
+}
+
+// Overload setting operator
+void Vector3d::operator =( Vector3d vNew ) {
+    updateVector( vNew[0], vNew[1], vNew[2] );
+}
+
+// Overload vector addition operator
+Vector3d Vector3d::operator +(Vector3d vec) {
+    Vector3d newVect;
+
+    for (int i=0; i<3; i++) {
+        newVect[i] = (*this)[i] + vec[i];
```

Comment:
What/where does (*this) points to/at?

Reply:
It's a c++ thing that refers to the specific instance of the object, similar to `self` in python. Jeff probably can explain it better if you're curious. Here, it just means that if I have a Vector3d + Vector3d, it adds them in the expected way.

Reply:
'this' is a pointer to the object - I assume here the Vector3d object (I don't remember the code)

Reply:
Oh - now I understand the question :-)  We're in an overload operator - 'this' points to the subect of the operator.  So here the '+' operator for the vector3d class is declared as:

Vector3d Vector3d::operator +(Vector3d vec)

so it would be used like a normal '+' in the code e.g. a = b + c, and
'this' refers to 'b', and 'vec' refers to 'c'  (and 'newVec' is what 'a' will get set to...